### ISSUE TITLE
[Feature]: sliced score matching for learning score functions.

### ISSUE THREAD

**Issue by tl82649** [gchq/coreax#75]
### Is your proposal related to a problem?

For the Stein kernel, we need to supply the score function $\nabla_\mathbf{x} \log f_X(\mathbf{x})$ of a PDF $f_X(\mathbf{x})$. For a finite sample of $n$ data points $(\mathbf{x}_1,\dots,\mathbf{x}_n)$ in $d$ dimensions, the score function needs to be inferred from the sample.

This is currently performed explicitly by fitting a kernel density estimate (KDE) to the data points, and computing the log transform and gradient of the induced density function; prior to use in the Stein kernel. This process involves the $\mathcal{O}(dn^2)$ KDE process, along with a choice of kernel function and hyperparameters (currently radial basis function with median heuristic, which also pointlessly computes the normalising constant). This could also result in model mismatch for complex, high-dimensional data distributions.

### Describe the solution you would like*

I propose using score matching to learn $\nabla_\mathbf{x} \log f_X(\mathbf{x})$ directly from $(\mathbf{x}_1,\dots,\mathbf{x}_n)$ using a neural network $h\_\theta(\mathbf{x})$ parameterised by weights $\theta$. Specifically, I intend to implement sliced score sampling as [outlined in Song _et al._'s 2019 paper.](https://arxiv.org/pdf/1905.07088.pdf)

Given the non-trivial function approximation properties of neural networks, this should allow for more complex score functions that better match those induced by the data distributions. The hope is that this will improve the coresets output by methods using the Stein kernel. Moreover, when learned with gradient-based optimisation, this should allow us to learn the score function in $\mathcal{O}(dn)$ time.

### Describe alternatives you've considered

Score matching originates from [Hyvärinen's 2005 paper](https://www.jmlr.org/papers/volume6/hyvarinen05a/hyvarinen05a.pdf). However, as [Song _et al_.](https://arxiv.org/pdf/1905.07088.pdf) point out, the Fisher divergence objective function involves computation of a Hessian trace, which is inefficient for large $d$ and not conducive to autodiff. 

Sliced score matching uses random vector projection to convert this to a Hessian-vector product (with optional neural network approximation of $\nabla_\mathbf{x} \log f_X(\mathbf{x})$ ), which can be computed efficiently in Jax.

### Code of Conduct*

- [X] I agree to follow this project's Code of Conduct
